bokehjs/src/lib/core/visuals.ts

```diff
@@ -8,6 +8,179 @@ import {LineJoin, LineCap, FontStyle, TextAlign, TextBaseline} from "./enums"
 import {HasProps} from "./has_props"
 import {ColumnarDataSource} from "models/sources/columnar_data_source"
 
+function _horz(ctx: Context2d , h: number, h2: number): void {
+  ctx.moveTo(0, h2+0.5)
+  ctx.lineTo(h, h2+0.5)
+  ctx.stroke()
+}
+
+function _vert(ctx: Context2d , h: number, h2: number): void {
+  ctx.moveTo(h2+0.5, 0)
+  ctx.lineTo(h2+0.5, h)
+  ctx.stroke()
+}
+
+function _x(ctx: Context2d , h: number): void {
+  ctx.moveTo(0, h)
+  ctx.lineTo(h, 0)
+  ctx.stroke()
+  ctx.moveTo(0, 0)
+  ctx.lineTo(h, h)
+  ctx.stroke()
+}
+
+function _get_canvas(size: number): HTMLCanvasElement {
+  const canvas = document.createElement('canvas')
+  canvas.width = size
+  canvas.height = size
+  return canvas
+}
+
+function create_hatch_canvas(hatch_pattern: string, hatch_color: string, hatch_scale: number, hatch_weight:number ): HTMLCanvasElement {
+    const h = hatch_scale
+    const h2 = h / 2
+    const h4 = h2 / 2
+
+    const canvas = _get_canvas(hatch_scale)
+
+    const ctx = canvas.getContext("2d")! as Context2d
```

Comment:
So my suggestion to use `Context2d` wasn't entirely correct, because this and what we use primarily for rendering are different types of contexts. I'm a bit surprised this type checks at all. Currently to assert that `ctx` is actually `Context2d`, is to call `fixup_ctx(ctx)` after obtaining the context. In future work I will abstract canvas rendering, so that we won't be dealing with native APIs directly, so there will be no confusion here.

Reply:
Just to be clear, you think these should be reverted to `CanvasRenderingContext2D` ?

Reply:
Not really, because I would like to have consistency here and I would prefer to use our refined APIs over native ones. I wouldn't be too concerned with this, as I will be cleaning up rendering anyway in the upcoming work.

Reply:
Ok, any objections if I merge this now as-is to get in a dev build?

Reply:
Actually it looks like there are some very slight image diffs for bars, I will look at this tonight

https://ci.bokeh.org/travis/525789333/examples.html

Reply:
Yeah, let's merge.

Reply:
OK, I will make a note to check out the diffs. It may actually be fine